### PR TITLE
feat(e2ei): respect e2ei during login and mls client creation (WPB-5851)

### PR DESCRIPTION
**Pull request by mchenani**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5851" title="WPB-5851" target="_blank">TaskWPB-5851</a>  [Android] Respect E2EI when Creating New MLS Client
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Respecting E2EI state during MLS client registration and login.

#### Test Coverage (Optional)
WIP
- [ ] I have added automated test to this contribution

#### How to Test

Have a team that has e2ei enabled, during login the user required to enroll certificate first.

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [2723](https://10.10.124.134/job/AR-build-pipeline/2723/) **failed**.

File: app/src/main/kotlin/com/wire/android/feature/e2ei/OAuthUseCase.kt [hunk 1]
```diff
@@ -177,6 +180,17 @@ class OAuthUseCase(context: Context, private val authUrl: String, oAuthState: St
         AuthorizationRequest.Scope.EMAIL,
         AuthorizationRequest.Scope.PROFILE,
         AuthorizationRequest.Scope.OFFLINE_ACCESS
+    ).setClaims(
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/feature/e2ei/OAuthUseCase.kt]: todo: will extract it in another PR.

File: app/src/main/kotlin/com/wire/android/migration/feature/MigrateClientsDataUseCase.kt [hunk 2]
```diff
@@ -103,6 +103,18 @@ class MigrateClientsDataUseCase @Inject constructor(
...
+                    is RegisterClientResult.E2EICertificateRequired->
+                        withTimeoutOrNull(SYNC_START_TIMEOUT) {
+                            syncManager.waitUntilStartedOrFailure()
+                        }.let {
+                            it ?: Either.Left(NetworkFailure.NoNetworkConnection(null))
+                        }.flatMap {
+                            syncManager.waitUntilLiveOrFailure()
+                                .onSuccess {
+                                    userDataStoreProvider.getOrCreate(userId).setInitialSyncCompleted()
+                                    TODO() //TODO: ask question about this!
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/migration/feature/MigrateClientsDataUseCase.kt]: todo: will fix it

File: app/src/main/kotlin/com/wire/android/ui/authentication/create/code/CreateAccountCodeViewModel.kt [hunk 3]
```diff
@@ -187,6 +187,10 @@ class CreateAccountCodeViewModel @Inject constructor(
                     is RegisterClientResult.Success -> {
                         onSuccess()
                     }
+                    is RegisterClientResult.E2EICertificateRequired ->{
+                        //todo
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/ui/authentication/create/code/CreateAccountCodeViewModel.kt]: todo: will cover it today

File: app/src/main/kotlin/com/wire/android/ui/settings/devices/e2ei/E2eiCertificateDetailsScreen.kt [hunk 4]
```diff
@@ -154,4 +154,4 @@ fun E2eiCertificateDetailsContent(
     )
 }
 
-const val CERTIFICATE_FILE_NAME = "certificate.pem"
+const val CERTIFICATE_FILE_NAME = "certificate.txt"
```
Comment (mchenani) [inline app/src/main/kotlin/com/wire/android/ui/settings/devices/e2ei/E2eiCertificateDetailsScreen.kt:157]: todo: will extract it in another PR.

**Comment by AndroidBob** [comment]
Build [2728](https://10.10.124.134/job/AR-build-pipeline/2728/) **failed**.

**Comment by AndroidBob** [comment]
Build [2730](https://10.10.124.134/job/AR-build-pipeline/2730/) **failed**.

**Comment by AndroidBob** [comment]
Build [2732](https://10.10.124.134/job/AR-build-pipeline/2732/) **failed**.

**Comment by AndroidBob** [comment]
Build [2740](https://10.10.124.134/job/AR-build-pipeline/2740/) **failed**.

**Comment by AndroidBob** [comment]
Build [2758](https://10.10.124.134/job/AR-build-pipeline/2758/) **failed**.

**Comment by AndroidBob** [comment]
Build [2760](https://10.10.124.134/job/AR-build-pipeline/2760/) **failed**.

**Comment by AndroidBob** [comment]
Build [2762](https://10.10.124.134/job/AR-build-pipeline/2762/) **failed**.

**Comment by AndroidBob** [comment]
Build [2768](https://10.10.124.134/job/AR-build-pipeline/2768/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(8730160)(8730160)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(8730160)(8730160)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt [hunk 5]
```diff
@@ -160,6 +166,18 @@ class WireActivityViewModel @Inject constructor(
...
+            observeUserId
+                .flatMapLatest {
+                    it?.let { observeIfE2EIRequiredDuringLoginUseCaseProviderFactory.create(it).observeIfE2EIIsRequiredDuringLogin() }
+                        ?: flowOf(null)
+                }
+                .distinctUntilChanged()
+                .collect { _observeE2EIState.emit(it) }
+        }
+    }
+
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:180]: instead of having 2 `val`s `_observeE2EIState` and `observeE2EIState` could be just one:
```
private val observeE2EIState = observeUserId
    .flatMapLatest {
        it?.let { observeIfE2EIRequiredDuringLoginUseCaseProviderFactory.create(it).observeIfE2EIIsRequiredDuringLogin() }
            ?: flowOf(null)
    }
    .distinctUntilChanged()
```

**Review by borichellow — COMMENTED** [review]
nice one! Just 1 small suggestion

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt (hunk 5 above)
Comment (mchenani) [reply app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:180]: @borichellow thanks, let's have this change in another pr to merge it later, to not disturb this PR.

**Review by borichellow — APPROVED** [review]

File: build-logic/plugins/src/main/kotlin/AndroidCoordinates.kt [hunk 6]
```diff
@@ -25,6 +25,6 @@ object AndroidSdk {
 
 object AndroidApp {
     const val id = "com.wire.android"
-    const val versionName = "4.6.0"
+    const val versionName = "4.7.0"
```
Comment (ohassine) [inline build-logic/plugins/src/main/kotlin/AndroidCoordinates.kt:28]: We shouldn't change the version

File: app/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -238,6 +238,15 @@
...
                 tools:node="remove" />
 
+        <!--
+        We make a custom Firebase init, because the app selects the Firebase project at a runtime,
+        so we need to remove default FirebaseInitProvider and create our custom FirebaseInitializer.
+        -->
+        <provider
+                android:name="com.google.firebase.provider.FirebaseInitProvider"
+                android:authorities="${applicationId}.firebaseinitprovider"
+                tools:node="remove" />
```
Comment (ohassine) [inline app/src/main/AndroidManifest.xml:248]: Do we need this in RC ?

**Review by ohassine — APPROVED** [review]
